### ISSUE TITLE
Route based on the url field.

### ISSUE THREAD

**Issue by pcoves** [jaspervdj/hakyll#819]
Hi there.

I'm working on an easy to use internationalization/i18n library that works with Hakyll.

I wrote a custom `Compiler` I use to fetch translations and I can now translate pieces of templates using a `functionField`.
I also wrote a custom `urlField` that translate directories and file names on the fly following the same principle.

But now, I'm looking for a way to use my i18n system when generating my Items route.
I saw https://jaspervdj.be/hakyll/reference/Hakyll-Core-Routes.html#v:metadataRoute but the `url` is not part of the metadata.

TL;DR : looking for a way to call `loadBody` in a `customRoute`.

Many thanks in advance.
> P.

**Comment by Minoru** [comment]
"Metadata" is what the front matter contains, i.e. this block at the beginning of the file:

```
title: …
author: …
published: false
```

The `url` comes from the context. [`Hakyll.Web.Template.Context.urlField`](https://jaspervdj.be/hakyll/reference/Hakyll-Web-Template-Context.html#v:urlField) actually examines the route to generate the URL, so that approach is barren.

Can you explain your i18n routing algorithm in a bit more detail? What information does it need, exactly?

**Comment by pcoves** [comment]
@Minoru Hey, thanks for your reply.

The I18n code is here `https://gitlab.com/swi18ng/swi18ng/-/blob/dev/src/I18n.hs` and use in `https://gitlab.com/swi18ng/swi18ng/-/blob/dev/src/Main.hs`.

In the first file at [line 80](https://gitlab.com/swi18ng/swi18ng/-/blob/dev/src/I18n.hs#L80) you can find my take at the `urlField`.
Its purpose is to split the paths and translate every part of it.
It works like a charm.

My question is more about how to use the computed `url` in the `route` part of my `Rules`.

**Comment by Minoru** [comment]
Oh, actual code is the best! Thanks for the links.

[On line 87](https://gitlab.com/swi18ng/swi18ng/-/blob/164b18a5b323686694689c2208bacd73b6bcd505/src/I18n.hs#L87), you call `getRoute`, which means `urlField` depends on the route. If you somehow made the route depend on the `url` field, you'd get a cycle. This is a dead end.

The only way out that I can think of is:
1. load all the possible translations before doing anything else (you can take inspiration from the [`templateCompiler`](https://jaspervdj.be/hakyll/reference/Hakyll-Web-Template.html#v:templateCompiler));
2. use [`customRoute`](https://jaspervdj.be/hakyll/reference/Hakyll-Core-Routes.html#v:customRoute) to generate the route from the identifier just like you currently do in `urlField`. This should use pre-loaded translations rather than call `loadBody` (which you can't call inside `Rules` anyway);
3. use the standard `urlField` which just looks at the route.

Does that make sense?

**Comment by pcoves** [comment]
That's precisely what I did at first : [loading all translations](https://gitlab.com/swi18ng/swi18ng/-/blob/master/src/Rules.hs#L37) and [carry them around](https://gitlab.com/swi18ng/swi18ng/-/blob/master/src/Main.hs#L15) in every `Rules`.

It works (I even use it to detemine what language to generate or not instead of hard coding the list), but there are some caveats.
My main concern is that it bypasses the dependency system built in Hakyll (updating a translation does not update the depending files).
Also, it does not feel *native* at all but some might accommodate with that.

Your explanation makes perfect sens and I prefer to build my package on top of Hakyll's dependency system.
I'll keep my `URL` untouched for now.

Thanks for you time.
I'll try to push this to a `hakyll-contrib-i18n` on hackage as it works perfectly well.

**Comment by Minoru** [comment]
Wait, turns out I was partially wrong: you *can* run `loadBody` in `Rules`, with `Haskyll.Core.Rules.compile . loadBody`! This doesn't fix the cycle, of course, but it opens up a way to fix the dependency tracking.

The crucial difference between your existing implementation and `templateCompiler` is that `templateCompiler` stores templates as `Item`s. Other items can then depend on templates using `loadTemplate`. If you do the same, and use `comile.loadBody` inside `Rules`, you should be able to access the translations from everywhere *and* get dependency tracking too.

**Comment by pcoves** [comment]
Yep. I do it [here](https://gitlab.com/swi18ng/swi18ng/-/blob/dev/src/Main.hs#L47) but it's in the `compile` function, not in the `route` one.

To use it with the `route` I'd have to do something like `route $ customRoute foo` with `foo :: Identifier -> FilePath`.
I'm not able to call `loadBody` in here.

**Comment by pcoves** [comment]
Also note that there won't be a dependency cycle if I simple set the `route`.

I would not have to toy around with a custom `urlField` at it's based on the `route`.

**Comment by Minoru** [comment]
No, no, I'm talking about `Rules`, not `Route`s. I mean something like this:

```haskell
match "posts/*.md" $ do
  translation <- compile . loadBody "tempaltes/translation.yml"
  route $ customRoute (routeWithThanslation translation)
  compile $ ...

  where
    routeWithTranslations :: Translation -> (Identifier -> FilePath)
    routeWithTranslations translation = {- a logic similar to your existing `urlField` -}
```

**Comment by pcoves** [comment]
But, `compile :: (Binary a, Typeable a, Writable a) => Compiler (Item a) -> Rules ()` so `compile . loadBody "tempaltes/translation.yml" :: ()` right?

In that case, `translation <- compile . loadBody "tempaltes/translation.yml"` does not give me anything I can use right?

**Comment by Minoru** [comment]
> so compile . loadBody "tempaltes/translation.yml" :: () right?

`… :: Rules ()`, I think.

> In that case, translation <- compile . loadBody "tempaltes/translation.yml" does not give me anything I can use right?

Yep, you're absolutely correct! Sorry, looks like I was wasting your time all along :(

**Comment by pcoves** [comment]
> … :: Rules (), I think.

Yep, my bad.

---

No worries, it's my first time working on a *real world problem* and I learned a lot writing the code and all that goes with it.

Untranslated URLs are fine to be honest and every thing else works just like I intended to so the only thing left is pushing this to hackage.

Thanks for you time!
